llpc/util/llpcElfWriter.cpp

```diff
@@ -1243,62 +1251,76 @@ void fixUpRelocations(ElfWriter<Elf> *writer, const std::vector<RelocationEntry>
+  ElfSectionBuffer<typename Elf::SectionHeader> *vertexShaderTextSection = nullptr;
+  relocatableElfs[0]->getTextSectionData(&vertexShaderTextSection);
+  ElfSectionBuffer<typename Elf::SectionHeader> *fragmentShaderTextSection = nullptr;
+  relocatableElfs[1]->getTextSectionData(&fragmentShaderTextSection);
+  ElfSectionBuffer<typename Elf::SectionHeader> *fetchShaderTextSection = nullptr;
+  relocatableElfs[2]->getTextSectionData(&fetchShaderTextSection);
+
+  // First merge the fetch shader and the vertex shader, and then merge with the fragment shader.
+  ElfSectionBuffer<typename Elf::SectionHeader> fullVertexShaderTextSection;
+  mergeSection(fetchShaderTextSection, alignTo(fetchShaderTextSection->secHead.sh_size, vertexShaderAlignment), nullptr,
+               vertexShaderTextSection, 0, nullptr, &fullVertexShaderTextSection);
+  mergeSection(&fullVertexShaderTextSection,
+               alignTo(fullVertexShaderTextSection.secHead.sh_size, fragmentShaderAlignment), nullptr,
+               fragmentShaderTextSection, 0, nullptr, &m_sections[m_textSecIdx]);
+  delete[] fullVertexShaderTextSection.data;
+
+  // We do not copy the fetch shader's tring table, so we need to make sure we use the string table offset for the name
+  // from the vertex shader.
+  m_sections[m_textSecIdx].secHead.sh_name = vertexShaderTextSection->secHead.sh_name;
 
   // Build the symbol table.  First set the symbol table section header.
   ElfSectionBuffer<typename Elf::SectionHeader> *symbolTableSection = nullptr;
   relocatableElfs[0]->getSectionDataBySectionIndex(relocatableElfs[0]->getSymSecIdx(), &symbolTableSection);
   m_sections[m_symSecIdx].secHead = symbolTableSection->secHead;
 
-  // Now get the symbols that belong in the symbol table.
-  unsigned offset = 0;
   std::vector<RelocationEntry> relocations;
 
   // Insert a dummy symbol. The ELF spec requires the symbol table to begin with a dummy symbol.
   getSymbol("")->secIdx = 0;
 
-  for (const ElfReader<Elf> *elf : relocatableElfs) {
+  // Now get the symbols that belong in the symbol table.  No symbols from the fetch shader are needed.
+  unsigned offset = alignTo(fetchShaderTextSection->secHead.sh_size, vertexShaderAlignment);
+  for (unsigned i = 0; i < 2; ++i) {
+    const ElfReader<Elf> *elf = relocatableElfs[i];
     unsigned relocElfTextSectionId = elf->GetSectionIndex(TextName);
     std::vector<ElfSymbol> symbols;
     elf->GetSymbolsBySectionIndex(relocElfTextSectionId, symbols);
     for (auto sym : symbols) {
       if (strncmp(sym.pSymName, "BB", 2) == 0) {
         continue;
       }
-      // When relocations start being used, we will have to filter out symbols related to relocations.
-      // We should be able to filter the BB* symbols as well.
       ElfSymbol *newSym = getSymbol(sym.pSymName);
       newSym->secIdx = m_textSecIdx;
       newSym->secName = nullptr;
-      newSym->value = sym.value + offset;
+      newSym->value = sym.value + (i == 0 ? 0 : offset);
```

Comment:
This seems wrong to me. Since i == 0 is the vertex shader, shouldn't it be offset by the (non-zero) vertex shader part base, and i == 1 should be offset by the fragment shader part base?

Maybe this could be cleaned up a little by having mergeSection take an ArrayRef of parts and output a vector of offsets for each part?

Ah, or is this because _amdgpu_vs_main wants to be special-cased? But in that case, it'd feel cleaner to me to explicitly special-case it by just setting its value below.

Reply:
Yes, I want to special case the vertex shader.  It should start at the beginning of the fetch shader, which will be included in the vertex shader, not at the location it is being copied.  I've fixed it up as you suggested.